### PR TITLE
Create workload permission objects per instance of ArgoCD CR

### PR DESCRIPTION
**Pull request by sbose78**
Broke away from https://github.com/argoproj-labs/argocd-operator/pull/161

### **Why**

The roles/rolebindings/service accounts responsible for powering an ArgoCD instance were not not being created per instance of ArgoCD CR. This prevented the user from being able to install the ArgoCD control plane in namespaces where the operator workload itself wasn't deployed.

With this change, you could create an `ArgoCD` CR in any namespace and get yourself an ArgoCD instance!
 

### **What**

- All roles, rolebindings, clusterrolebindings and SAs associated with the control plane objects would be owned by the ArgoCD CR

- Only the "read all" cluster role for server and application-controller would be created/managed outside the operator ( by OLM ) since they wouldn't necessarily differ per installation of argocd

### **Next PR**
* Release ArgoCD in all-namespaces mode.

### **Future PRs**
Support setting up of cluster config ClusterRole

### CODE REVIEW

File: pkg/common/annotations.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package common
+
+const (
+	// AnnotationName is the annotation on child resources that specifies which ArgoCD instance
+	// name a specific object is associated with
+	AnnotationName = "argocds.argoproj.io/name"
```
Comment (wtam2018) [inline pkg/common/annotations.go:6]: Using the plural form of argocd (argocds) seems odd to me.   Is it intentional?

Comment (sbose78) [reply pkg/common/annotations.go:6]: The CRD is already named that way :) 

```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: argocds.argoproj.io
spec:
  group: argoproj.io
  names:
    kind: ArgoCD
    listKind: ArgoCDList
    plural: argocds
```

File: pkg/controller/argocd/util.go [hunk 2]
```diff
@@ -639,14 +640,23 @@ func labelsForCluster(cr *argoprojv1a1.ArgoCD) map[string]string {
 	return labels
 }
 
+// annotationsForCluster returns the labels for all cluster resources.
+func annotationsForCluster(cr *argoprojv1a1.ArgoCD) map[string]string {
+	labels := argoutil.DefaultAnnotations(cr)
+	for key, val := range cr.ObjectMeta.Annotations {
+		labels[key] = val
+	}
+	return labels
```
Comment (wtam2018) [inline pkg/controller/argocd/util.go:649]:
```suggestion
// annotationsForCluster returns the annotations for all cluster resources.
func annotationsForCluster(cr *argoprojv1a1.ArgoCD) map[string]string {
	annotations := argoutil.DefaultAnnotations(cr)
	for key, val := range cr.ObjectMeta.Annotations {
		annotations[key] = val
	}
	return annotations
```

**Comment by wtam2018** [comment]
/lgtm

@sbose78 Note: this is a conflicting file.

**Comment by wtam2018** [comment]
@sbose78 please rebase.


Conflicting files
pkg/controller/argocd/util.go

**Comment by sbose78** [comment]
Sure William!

File: pkg/controller/argocd/rolebinding.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	roleBinding := newRoleBindingWithname(name, cr)
+
+	// fetch existing rolebinding by name
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: roleBinding.Name, Namespace: cr.Namespace}, roleBinding)
+	roleBindingExists := true
+	if err != nil {
+		if errors.IsNotFound(err) {
+			roleBindingExists = false
+			roleBinding = newRoleBindingWithname(name, cr)
+		} else {
```
Comment (bigkevmcd) [inline pkg/controller/argocd/rolebinding.go:76]: You could invert this code and drop the `else`.

File: pkg/controller/argocd/rolebinding.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	roleBinding.RoleRef = v1.RoleRef{
+		APIGroup: v1.GroupName,
+		Kind:     "Role",
+		Name:     role.Name,
+	}
+
+	controllerutil.SetControllerReference(cr, roleBinding, r.scheme)
+	if roleBindingExists {
+		err = r.client.Update(context.TODO(), roleBinding)
```
Comment (bigkevmcd) [inline pkg/controller/argocd/rolebinding.go:96]: You could simplify this to: (and there's another case below with the same idea).

```Go
if roleBindingExists {
		return r.client.Update(context.TODO(), roleBinding)
}
return r.client.Create(context.TODO(), roleBinding)
```

File: pkg/controller/argocd/rolebinding.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+	// fetch existing rolebinding by name
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: roleBinding.Name}, roleBinding)
+	roleBindingExists := true
+	if err != nil {
+		if errors.IsNotFound(err) {
+			roleBindingExists = false
+			roleBinding = newClusterRoleBindingWithname(name, cr)
+		} else {
+			return err
```
Comment (bigkevmcd) [inline pkg/controller/argocd/rolebinding.go:116]: See above for the same issue, inversion would simplify the code.

File: pkg/controller/argocd/rolebinding_test.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+func TestReconcileArgoCD_reconcileClusterRoleBinding(t *testing.T) {
+	logf.SetLogger(logf.ZapLogger(true))
+	a := makeTestArgoCD()
+	r := makeTestReconciler(t, a)
+
+	assertNoError(t, r.reconcileClusterRoleBinding("x", &v1.ClusterRole{}, &corev1.ServiceAccount{}, a))
+
+	// TODO: Compare contents of the object
```
Comment (bigkevmcd) [inline pkg/controller/argocd/rolebinding_test.go:28]: This bit feels like it still needs to be done?

File: pkg/controller/argocd/service_account.go [hunk 7]
```diff
@@ -95,3 +125,61 @@ func (r *ReconcileArgoCD) reconcileDexServiceAccount(cr *argoprojv1a1.ArgoCD) er
 
 	return r.client.Update(context.TODO(), sa)
 }
+
+func (r *ReconcileArgoCD) reconcileServiceAccountClusterPermissions(name string, cr *argoprojv1a1.ArgoCD) error {
+	var role *v1.ClusterRole
+	var sa *corev1.ServiceAccount
+	var err error
+
+	// setup role for
```
Comment (bigkevmcd) [inline pkg/controller/argocd/service_account.go:134]: This comment is incomplete?

File: pkg/controller/argocd/service_account.go [hunk 8]
```diff
@@ -95,3 +125,61 @@ func (r *ReconcileArgoCD) reconcileDexServiceAccount(cr *argoprojv1a1.ArgoCD) er
...
+
+	return r.reconcileClusterRoleBinding(name, role, sa, cr)
+}
+
+func (r *ReconcileArgoCD) reconcileServiceAccountPermissions(name string, rules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) error {
+	var role *v1.Role
+	var sa *corev1.ServiceAccount
+	var err error
+
+	// setup role for
```
Comment (bigkevmcd) [inline pkg/controller/argocd/service_account.go:152]: Copy paste from above, with the same incomplete comment?

File: pkg/controller/argocd/service_account_test.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+func TestReconcileArgoCD_reconcileServiceAccountPermissions(t *testing.T) {
+	logf.SetLogger(logf.ZapLogger(true))
+	a := makeTestArgoCD()
+	r := makeTestReconciler(t, a)
+	_, err := r.reconcileRole("x", []v1.PolicyRule{}, a)
+	assertNoError(t, err)
+
+	assertNoError(t, r.reconcileServiceAccountPermissions("x", []v1.PolicyRule{}, a))
+}
```
Comment (bigkevmcd) [inline pkg/controller/argocd/service_account_test.go:32]: Shouldn't this make some assertions about what was reconciled?

Also, let's ensure that if it's rereconciled, it continues to work?

**Review by bigkevmcd — CHANGES_REQUESTED** [review]
There are a number of places where bare `err` is returned, which could possibly use some explanation of the failure `fmt.Errorf`.

**Review by chetan-rns — COMMENTED** [review]
LGTM. Tried it locally and was able to create multiple instances in different namespaces

**Review by chetan-rns — APPROVED** [review]

**Pull request merged**